### PR TITLE
Server side changes for video calls

### PR DESCRIPTION
**Pull request by aforge**
Calls are initiated by publishing a {data} with head "application/x-tinode-webrtc" (`handleCallInvite()` called from `handlePubBroadcast()`).
Later call events/data are exchanged via {note} & {info} pairs (`handleCallEvent()` called from `handleNoteBroadcast()`).

### CODE REVIEW

File: server/topic.go [hunk 1]
```diff
@@ -976,44 +1093,261 @@ func (t *Topic) handlePubBroadcast(msg *ClientComMessage) {
...
-		sess:      msg.sess,
+	if t.isReadOnly() {
+		msg.sess.queueOut(ErrPermissionDenied(msg.Id, t.original(asUid), msg.Timestamp))
+		return
 	}
-	if msg.Pub.NoEcho {
-		data.SkipSid = msg.sess.sid
+
+	//
+	var params map[string]interface{} = nil
```
Comment (or-else) [inline server/topic.go:1116]: no need for ` = nil`

File: server/topic.go [hunk 2]
```diff
@@ -976,44 +1093,261 @@ func (t *Topic) handlePubBroadcast(msg *ClientComMessage) {
...
+	isCall := msg.Pub.Head != nil && msg.Pub.Head["mime"] == constTinodeVideoCallMimeType
+	if isCall {
+		if t.currentCall != nil {
+			msg.sess.queueOut(ErrCallBusyReply(msg, types.TimeNow()))
+			return
+		}
+		params = make(map[string]interface{})
```
Comment (or-else) [inline server/topic.go:1123]: remove

File: server/topic.go [hunk 3]
```diff
@@ -976,44 +1093,261 @@ func (t *Topic) handlePubBroadcast(msg *ClientComMessage) {
...
+		params["sip-code"] = 100
```
Comment (or-else) [inline server/topic.go:1124]: `params = map[string]interface{}{"sip-code": 100}`

Comment (or-else) [reply server/topic.go:1124]: Why 100? Magic constant?

File: server/topic.go [hunk 4]
```diff
@@ -976,44 +1093,261 @@ func (t *Topic) handlePubBroadcast(msg *ClientComMessage) {
...
-	t.broadcastToSessions(data)
+// Ends current call in response to a client hangup request (msg).
+func (t *Topic) maybeEndCallInProgress(from string, msg *ClientComMessage) {
+	if t.currentCall == nil {
+		return
+	}
+	t.callEstablishmentTimer.Stop()
+	originator, _ := t.getCallOriginator()
+	var replaceWith string
+	if len(from) > 0 && len(t.currentCall.parties) == 2 {
```
Comment (or-else) [inline server/topic.go:1150]: `len(from) > 0` -> `from != ""`

File: server/topic.go [hunk 5]
```diff
@@ -976,44 +1093,261 @@ func (t *Topic) handlePubBroadcast(msg *ClientComMessage) {
...
+		replaceWith = constCallMsgFinished
+	} else {
+		// Call hasn't been established. Just drop it.
+		replaceWith = constCallMsgDisconnected
+	}
 
-	// usersPush will update unread message count and send push notification.
-	if pushRcpt := t.pushForData(asUid, data.Data); pushRcpt != nil {
-		usersPush(pushRcpt)
+	// Send message message indicating the call has ended.
```
Comment (or-else) [inline server/topic.go:1158]: message message -> message

File: server/datamodel.go [hunk 6]
```diff
@@ -1548,6 +1560,26 @@ func ErrPolicyReply(msg *ClientComMessage, ts time.Time) *ServerComMessage {
 	return ErrPolicyExplicitTs(msg.Id, msg.Original, ts, msg.Timestamp)
 }
 
+// ErrCallBusyExplicitTs indicates a "busy" reply to a video call request (486).
+func ErrCallBusyExplicitTs(id, topic string, serverTs, incomingReqTs time.Time) *ServerComMessage {
+	return &ServerComMessage{
+		Ctrl: &MsgServerCtrl{
+			Id:        id,
+			Code:      486, // Busy here.
+			Text:      "Busy Here",
```
Comment (or-else) [inline server/datamodel.go:1569]: All other `Text` messages are lowercase. Keep it that way?

File: server/topic.go [hunk 7]
```diff
@@ -923,49 +1008,81 @@ func (t *Topic) sendSubNotifications(asUid types.Uid, sid, userAgent string) {
...
 
-// handlePubBroadcast fans out {pub} -> {data} messages to recipients in a master topic.
-// This is a NON-proxy broadcast.
-func (t *Topic) handlePubBroadcast(msg *ClientComMessage) {
-	asUid := types.ParseUserId(msg.AsUser)
-	if t.isInactive() {
-		// Ignore broadcast - topic is paused or being deleted.
-		msg.sess.queueOut(ErrLocked(msg.Id, t.original(asUid), msg.Timestamp))
-		return
+func (t *Topic) broadcastMessage(msg *ClientComMessage, asUid types.Uid, noEcho bool, replyParams, head map[string]interface{}, content interface{}) {
```
Comment (or-else) [inline server/topic.go:1011]: Comment, please.

File: server/topic.go [hunk 8]
```diff
@@ -976,44 +1093,261 @@ func (t *Topic) handlePubBroadcast(msg *ClientComMessage) {
...
-		Timestamp: msg.Timestamp,
-		sess:      msg.sess,
+	if t.isReadOnly() {
+		msg.sess.queueOut(ErrPermissionDenied(msg.Id, t.original(asUid), msg.Timestamp))
+		return
 	}
-	if msg.Pub.NoEcho {
-		data.SkipSid = msg.sess.sid
+
+	//
```
Comment (or-else) [inline server/topic.go:1115]: Remove

File: server/pres.go [hunk 9]
```diff
@@ -498,6 +499,32 @@ func (t *Topic) infoSubsOffline(from types.Uid, what string, seq int, skipSid st
 	}
 }
 
+// Publish {info what=call} to topic subscribers's sessions on subscriber's 'me'.
+func (t *Topic) callSubsOffline(from string, target types.Uid, event string, seq int,
+                                sdp json.RawMessage, skipSid string, offlineOnly bool) {
+	if target.IsZero() {
+		logs.Err.Printf("callSubs could not find target: topic %s - from %s", t.name, from)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (aforge) [reply server/pres.go:508]: Done.

File: server/session.go [hunk 10]
```diff
@@ -1212,6 +1213,10 @@ func (s *Session) note(msg *ClientComMessage) {
 		if msg.Note.SeqId <= 0 {
 			return
 		}
+	case "call":
```
[the remark this answers is not in the document]
Comment (aforge) [reply server/session.go]: Done.

File: server/datamodel.go [hunk 11]
```diff
@@ -296,14 +296,19 @@ type MsgClientDel struct {
 
 // MsgClientNote is a client-generated notification for topic subscribers {note}.
 type MsgClientNote struct {
-	// There is no Id -- server will not akn {ping} packets, they are "fire and forget"
+	Id string `json:"id,omitempty"`
```
[the remark this answers is not in the document]
Comment (or-else) [reply server/datamodel.go:299]: > failures in those cases prevents the call from starting on the clients

The error handling there is just for reporting the failure to the user earlier. It does not prevent the failure: the call may still fail even if there is no error in sending the metadata. It still has to time out if it's not established in a reasonable time even if metadata was sent successfully.

File: server/session.go [hunk 12]
```diff
@@ -1259,7 +1269,7 @@ func (s *Session) expandTopicName(msg *ClientComMessage) (string, *ServerComMess
 		routeTo = msg.AsUser
 	} else if msg.Original == "fnd" {
 		routeTo = types.ParseUserId(msg.AsUser).FndName()
-	} else if strings.HasPrefix(msg.Original, "usr") {
+	} else if types.IsP2PTopic(msg.Original) {
```
Comment (or-else) [inline server/session.go:1272]: `types.GetTopicCat`

File: server/store/types/types.go [hunk 13]
```diff
@@ -219,6 +219,11 @@ func IsChannel(name string) bool {
 	return strings.HasPrefix(name, "chn")
 }
 
+// IsP2PTopic checks if the given topic name is a reference to a P2P topic.
+func IsP2PTopic(name string) bool {
```
Comment (or-else) [inline server/store/types/types.go:223]: `types.GetTopicCat`

**Comment by or-else** [comment]
Maybe move call-related stuff to a separate file?

File: server/session.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (or-else) [reply server/session.go]: I would do it differently:
```go
	case "call":
		if !types.IsP2PTopic(msg.Original) {
			// Calls are only available in P2P topics.
			return
		}
		fallthrough
	case "read", "recv":
		if msg.Note.SeqId <= 0 {
			return
		}
```

File: server/datamodel.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (aforge) [reply server/datamodel.go]: Okay. I have removed `Id`.

File: server/session.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (aforge) [reply server/session.go]: Done.

**Comment by aforge** [comment]
> Maybe move call-related stuff to a separate file?

Moved to `calls.go`.

File: server/datamodel.go [hunk 14]
```diff
@@ -296,14 +296,17 @@ type MsgClientDel struct {
 
 // MsgClientNote is a client-generated notification for topic subscribers {note}.
 type MsgClientNote struct {
-	// There is no Id -- server will not akn {ping} packets, they are "fire and forget"
```
Comment (or-else) [inline server/datamodel.go:299]: Please put back the comment.

**Review by or-else — APPROVED** [review]

File: server/datamodel.go (hunk 14 above)
Comment (aforge) [reply server/datamodel.go]: Done.

**Pull request merged**